### DISCUSSION TITLE
rebar on MSI Z97A Gaming 6 (7917)

### DISCUSSION THREAD

**Discussion by durdin85** [xCuri0/ReBarUEFI#44] [Q&A] [answered]
Hi, I was trying to add FFS module to the A.3 bios for this board [MSI Z97A Gaming 6](https://www.msi.com/Motherboard/Z97A-GAMING-6/support), but I am bit puzzled where to actually insert the ffs.
In the example screenshots, the section contains modules and then free space, but in my case, the section contains modules, then some padding and raw data:

![original-bios](https://user-images.githubusercontent.com/6603216/227639009-c5dd0ceb-bd23-4f7c-b20b-b8d9330cb331.png)

I've tried adding the rebar ffs, which then marks another sections for rebuild and when I open modified bios again, the padding is gone and there's only the free space:

![modified-bios](https://user-images.githubusercontent.com/6603216/227639919-c58d4f6b-83db-46d6-91bd-8e70c4dc2eca.png)

I am not sure what is the purpose of the raw module, but I suspect it might store some important data and be somewhere hard-referenced to fixed position on the end of the section. So flashing modified bios might be dangerous as there's no backup ROM available on this board and I do not have proper tools to reprogram the chip in case of bricked board.

So my question is if there's a way how to insert ffs without breaking the layout? I might try creating new shortened pad file (it seems it's just zeroes), but it also seems the original padding get's ignored, and I have very little experience with this toolset, so not sure how big the final padding file shall be?
Or do anyone have experience with this type of bios/board? I see some Z97 MSI boards as working in list of motherboards, so I wonder what I am doing wrong... or if I am missing something?

**Comment by xCuri0** [comment]
Have you tried adding it after the raw section ?

**Reply by durdin85** [reply]
Yes I've tried it as well, but it strips the pad section as well.
But I am just reading wiki and there's mentioned some pad file issue, so I'll give a try to MMTool.

**Reply by durdin85** [reply]
I've been able to insert the module using MMTool-4.50, and when I open the modified Bios image in uefitool order looks like the original except there's also the ReBarDxe inserted. But the uefitool shows it as last module before pad and the raw section, but in mmtool it's last module after the raw so I am still not 100% sure this will work.
Is there any way how to verify it except flashing?

**Comment by durdin85** [comment, marked as answer]
I've had some time to flash the bios and confirm that **MSI Z97A Gaming 6 (7917)** is working. But the Rebar module needs to be inserted via **MMTool 4.50**. Then order of modules and padding file is kept correctly. There's also hidden setting for Above 4G decoding in BIOS, but I was unable to get in visible in the menu using Aptiotools so this setting still needs to be enabled via [grub_mod_setup_var](https://github.com/xCuri0/ReBarUEFI/wiki/Enabling-hidden-4G-decoding).
I've been testing the BAR values up to 1G so far and it looks like it _works_ on this motherboard. I'll test bigger BAR sizes later on.
